### PR TITLE
Compare chain id env variable with rpc on startup

### PR DESCRIPTION
**Pull request by kimpers**
Implements #171

### CODE REVIEW

File: core/core.go [hunk 1]
```diff
@@ -711,6 +745,14 @@ func (app *App) Start(ctx context.Context) error {
 		}
 	}
 
+	// Need to wait for the async ETH RPC call to finish before we know if the chain id is correctly configured
+	chainIdCheckerWg.Wait()
+	if err := <-chainIdMismatchErrChan; err != nil {
```
Comment (fabioberger) [inline core/core.go]: Move this into a `case` statement in the above select block.

File: core/core.go [hunk 2]
```diff
@@ -552,6 +552,40 @@ func (app *App) Start(ctx context.Context) error {
...
 
+	// Ensure that RPC client is on the same ChainID as is configured with ETHEREUM_CHAIN_ID
+	chainIdCheckerWg := &sync.WaitGroup{}
+	chainIdMismatchErrChan := make(chan error, 1)
+	wg.Add(1)
+	chainIdCheckerWg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer chainIdCheckerWg.Done()
+		ctx, cancel := context.WithTimeout(context.Background(), ethereumRPCRequestTimeout)
```
Comment (fabioberger) [inline core/core.go]: Can we move the actual logic in this go routine into a helper method with a good name? Having it in-lined here feels like the wrong level of abstraction when compared with the rest of it's parent's function logic.

Comment (fabioberger) [reply core/core.go]: Use `innerCtx` instead of `context.Background()` here.

**Comment by albrow** [comment]
@kimpers @fabioberger do you mind if I take over this PR to push it over the finish line? This check would actually be really useful and solve a common problem, especially in the browser.

**Comment by kimpers** [comment]
> @kimpers @fabioberger do you mind if I take over this PR to push it over the finish line? This check would actually be really useful and solve a common problem, especially in the browser.

Feel free 👍 Beware that there seems to be a strange race condition with the test, locally on my machine it passes some times and fails other times. 

When you figure out what the underlying issue is I'd be very interested to know what causes it.

**Comment by albrow** [comment]
@kimpers the race condition here was definitely a tricky one. It doesn't directly stem from the code you added, moreso this new code made an existing bug more apparent.

Inside of `core.App.Start` there is a `select` statement where we wait for several different goroutines to send an error through their respective error channels (e.g. `p2pErrChan`, `orderWatcherErrChan`, or the new `chainIDMismatchErrChan` which you added in this PR):

```go
select {
		case err := <-p2pErrChan:
			if err != nil {
				log.WithError(err).Error("p2p node exited with error")
				cancel()
				return err
			}
		case err := <-orderWatcherErrChan:
			if err != nil {
				log.WithError(err).Error("order watcher exited with error")
				cancel()
				return err
			}

		// etc...
}
```

There was an implicit assumption here that fi one of the goroutines sent an error through their respective channels, it was time to shut the app down. But that assumption was wrong. There are some cases where a goroutine may send `nil` through the channel to indicate that it has exited _without_ an error, and in this case we don't shut down the application. The most straightforward example of this is that the ETH RPC rate limiter sends `nil` through `ethRPCRateLimiterErrChan` immediately if ETH RPC rate limiting has been disabled. In the old code this would cause us to exit the `select` statement, and therefore not receive errors from any of the other (still running) goroutines.

I fixed this by wrapping the `select` statement in a `for` loop. This way if one of the goroutines exits and sends `nil` through its respective error channel, we keep waiting to receive errors on the other channels and then handle them appropriately.

**Comment by albrow** [comment]
@fabio This PR is now ready for review. I changed it to no longer be a draft.

**Pull request merged**